### DISCUSSION TITLE
Re-referencing when one of the reference channels is not in the channels list

### DISCUSSION THREAD

**Discussion by stonekate** [bnicenboim/eeguana#191] [Q&A] [answered]
Hi Bruno, this is probably a silly question: I want to re-reference to the average of the two mastoids but I'm using a different BrainVision setup where the online reference (right mastoid = RM) does not appear in the Channels list of the .vhdr file. It's listed separately as "Reference Channel Name" (see line 101 of attached header file). When I load the eeg data, the online reference doesn't appear in `channel_names` or in the signal table. Have you come across this before?

[9.txt](https://github.com/bnicenboim/eeguana/files/8519296/9.txt)

**Comment by bnicenboim** [comment, marked as answer]
no, but you can simply do `data %>% eeg_mutate(RM = channel_dbl(0))`, since it's referenced to it, it must be zero.
But if you have a minimal example with the vhdr, vmrk and the dat files (preferably a short dat file), I can make it automatic for the next time.

It'll be useful to see other BV files anyway just in case.

**Reply by stonekate** [reply]
Cool, I figured making it 0 might be an option but wasn't sure. I put some data in Dropbox... I can't figure out how to make the dat file smaller :(

**Reply by stonekate** [reply]
Oh no I tried setting RM as a new zero channel but the data doesn't look right after re-referencing... maybe it's already re-referenced? But I didn't tell it to? Anyway, I'll have to look into the BV documentation tomorrow to see what it's doing.

Just using the online reference (not re-referencing):
![image](https://user-images.githubusercontent.com/12735809/164247426-de4cf354-85a0-47a5-98e9-bb2baa4be31a.png)


Re-referencing to average of LM and RM (RM as zero):
![image](https://user-images.githubusercontent.com/12735809/164247253-49f87116-b2b9-40b7-a23e-a30e58a4b8f8.png)

**Comment by bnicenboim** [comment]
oh, I don't know what's happening, but if the data is originally referenced to RM, then it means that you subtract RM from every channel, that means that RM channel must be zero. So that cannot be wrong.
maybe it's just the change of scale? HEOG and VEOG also look very different but should be changed (you shouldn't re-reference them)

**Comment by stonekate** [comment]
I think you're right, it's just the scale. Any tip on how to set it manually? I tried scale_y_continuous() and ylim() but nothing changed. 

I think the scale is why V/HEOG look weird (they're just really zoomed in, they weren't re-referenced).

**Reply by bnicenboim** [reply]
`coord_cartesian()`,  you should set   clip = "off" to have the signal going over their own channel, clip ="on" will cut stuff off.

**Reply by stonekate** [reply]
ok yeah that was it, now it looks as expected. thanks!

**Comment by spressi** [comment]
@bnicenboim would it be possible to add a convenience parameter to `eeg_rereference` for reinstating the implicit reference? It could look like `reinstate="Cz"` and maybe do some additional housekeeping like setting the channel number to "Ref_online" to mark it as the online reference during recording or set the resolution automatically.

I also noticed that `eegfile %>% read_vhdr() %>% channels_tbl()` has the `.reference` column set to `NA`. Would it be possible to read out the online reference (doesn't seem to be stored in the .vhdr or .vmrk file though) and enter it such that `reinstate=T` would restore the online reference channel? Not sure how this works for other file formats but reinstating online/implicit reference channels upon rereferencing is very common and I also stumbled across it :)

Keep up the very good work! :)

**Reply by bnicenboim** [reply]
Sure, open a feature request! Write a small piece of code showing how it will look like with the new functions/arguments and what would be the expected output at each step.

(Also feel free to take a look a the code, and see if after discuss how it will look like,  maybe you can even implement it, there's a branch called experimental, that is the most updated, but not tested enough version of the package)

I'm finally running a new EEG experiment, so I will start adding more features to the package again (but next year).
I decided to bite the python bullet, and start relying on MNE much more, this will hopefully speed up implementing new stuff (on the expense of potential installation issues).
